server/app/middlewares/websocket_backend.rb

```diff
@@ -67,29 +75,104 @@ def call(env)
     end
   end
 
-  ##
-  # On websocket connection open
-  #
-  # @param [Faye::WebSocket] ws
-  # @param [Rack::Request] req
-  def on_open(ws, req)
+  # @param node_id [String] request header Kontena-Node-Id
+  # @param grid_token [String] request header Kontena-Grid-Token
+  # @param init_attrs [Hash] initialize attributes on new node
+  # @raise [CloseError]
+  # @return [HostNode] with node_id set
+  def find_node_by_grid_token(node_id, grid_token, init_attrs)
     # check grid
-    grid = Grid.find_by(token: req.env['HTTP_KONTENA_GRID_TOKEN'].to_s)
-    unless grid
-      logger.error 'invalid grid token, closing connection'
-      ws.close(4001, "Invalid grid token")
-      return
+    grid = Grid.find_by(token: grid_token.to_s)
+
+    raise CloseError.new(4001), "Invalid grid token" unless grid
+
+    node = grid.host_nodes.find_by(node_id: node_id)
+
+    if !node
+      node = grid.host_nodes.create!(node_id: node_id, **init_attrs)
```

Comment:
There's a brief race condition window between the `grid.host_nodes.find_by(node_id: node_id)` and here, if two agents connect with the same node ID at the same time. With the unique index on the `node_id`, this will fail with a `Mongo::Error::OperationFailure`.

That error gets logged on the server, but it kills the `WebsocketBackend#on_open`, and the agent will presumably just be left hanging without seeing any error or response from the server :disappointed: